Parquet: Implement column index filter and update row read path to support page skipping

zhongyujiang opened the pull request:
This PR mainly does two things to support page filtering in the Parquet row read path:

1. This implements a `ParquetColumnIndexFilter` which is equivalent to the Parquet-mr's `ColumnIndexFilter`, but based on Iceberg's expressions. It uses reflection to call the `getColumnIndex` method in the `ParquetFileReader` to get the column index and passes it to the `ColumnIndexEvalVisitor` for `RowRanges` calculation. When filtering pages, it uses an `IntPredicate`  which takes a page index as an argument and evaluates whether the corresponding page can match the expression, using page stats stored in the column index.  Then geta a list of possible matching pages and geta `RowRanges` based on them.  This mocks column index stats for testing in the `TestColumnIndexFilter`, which follows the test method in the Parquet-mr project.

2. This also updates the Parquet row read path to support page skipping, it calculates `RowRanges` for row groups and makes`ParquetReader` to read the filtered page store (by using reflection)  if  the `RowRanges` is not empty and the row count of `RowRanges` < the row count in `BlockMetadata`.  The rows in the filtered page store may be not aligned between pages of the different columns, so this updates `ColumnIterator` to skip values which does not belong the target rows represented by `RowRanges`. 
In the page filtering scenario, the row reader will make its all column reader in sync before reading a new row. Specifically, the calculated `RowRanges` (target row indexes) will be passed to each `ColumnIterator`, and the `firstRowIndex` will be retrieved from pages for locating the initial value of current row index after reading a new page. This adds a `topLevel` field in `org.apache.iceberg.parquet.ParquetValueReaders.StructReader` to mark whether it is a row reader or nested struct  reader(only adds to Spark `InternalRowReader`), before the row `StructReader` reads a new row, its all child columns will skip un-target triples to sync. To achieve that, this adds `currentRowIndex` and `skipValues` in `ColumnIterator` to mark the state of synchronization, the `skipValues` represents the number of non-null values to skip. During the synchronization, the `currentRowIndex` will be incremented by 1 whenever rep level is 0 and the `skipValues` will be incremented by 1 whenever def level is is bigger than the `definitionLevel`,  synchronization will advance the `ColumnIterator` until the `currentRowIndex` is no longer less than `targetRowIndex`, and then call `PageIterator#skip` to skip values in the page. After the synchronization, all non-target triples before the next target row will be skipped, and the new row can be safely read.

The batch read path will be updated in a subsequent PR.
Coauthored by @rdblue

zhongyujiang commented:
@rdblue @Fokko Can you help to review this?

rdblue commented:
@zhongyujiang can you please add more to the description about what is included here and how you solved the problems with record materialization?

zhongyujiang commented:
@rdblue The description has been updated, I hope I made it clear, please let me know if something is unclear.

zhongyujiang commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetValueReader.java:
```diff
@@ -28,5 +30,5 @@
   List<TripleIterator<?>> columns();
 
-  void setPageSource(PageReadStore pageStore, long rowPosition);
+  void setPageSource(PageReadStore pageStore, long rowPosition, Optional<RowRanges> rowRanges);
```
>   java.method.numberOfParametersChanged: The number of parameters of the method have changed.
>   
>   old: method void org.apache.iceberg.parquet.ParquetReader<T>::<init>(org.apache.iceberg.io.InputFile, org.apache.iceberg.Schema, org.apache.parquet.ParquetReadOptions, java.util.function.Function<org.apache.parquet.schema.MessageType, org.apache.iceberg.parquet.ParquetValueReader<?>>, org.apache.iceberg.mapping.NameMapping, org.apache.iceberg.expressions.Expression, boolean, boolean)
>   new: method void org.apache.iceberg.parquet.ParquetReader<T>::<init>(org.apache.iceberg.io.InputFile, org.apache.iceberg.Schema, org.apache.parquet.ParquetReadOptions, java.util.function.Function<org.apache.parquet.schema.MessageType, org.apache.iceberg.parquet.ParquetValueReader<?>>, org.apache.iceberg.mapping.NameMapping, org.apache.iceberg.expressions.Expression, boolean, boolean, boolean)
>   
>   SOURCE: BREAKING, BINARY: BREAKING
>   
>   From old archive: iceberg-parquet-1.1.0.jar
>   From new archive: iceberg-parquet-3e8b750.jar
>   
>   If this is an acceptable break that will not harm your users, you can ignore it in future runs like so for:
>   
>     * Just this break:
>         ./gradlew :iceberg-parquet:revapiAcceptBreak --justification "{why this break is ok}" \
>           --code "java.method.numberOfParametersChanged" \
>           --old "method void org.apache.iceberg.parquet.ParquetReader<T>::<init>(org.apache.iceberg.io.InputFile, org.apache.iceberg.Schema, org.apache.parquet.ParquetReadOptions, java.util.function.Function<org.apache.parquet.schema.MessageType, org.apache.iceberg.parquet.ParquetValueReader<?>>, org.apache.iceberg.mapping.NameMapping, org.apache.iceberg.expressions.Expression, boolean, boolean)" \
>           --new "method void org.apache.iceberg.parquet.ParquetReader<T>::<init>(org.apache.iceberg.io.InputFile, org.apache.iceberg.Schema, org.apache.parquet.ParquetReadOptions, java.util.function.Function<org.apache.parquet.schema.MessageType, org.apache.iceberg.parquet.ParquetValueReader<?>>, org.apache.iceberg.mapping.NameMapping, org.apache.iceberg.expressions.Expression, boolean, boolean, boolean)"
>     * All breaks in this project:

The CI failed because some changes in this PR breaks the API compatibility check.  
Although these APIs are public, they are only used in the iceberg-parquet project, 
not used in engine specific readers. Can I expect the same behavior from those  external 
connectors not in this repo? So I can make the revapi check accept these changes. Or should 
I make this change by deprecating  and adding a new method?

hengqujushi commented:
How's it going?

sunchao commented:
@zhongyujiang @rdblue any recent progress on this PR? it would a very useful feature!

zhongyujiang commented:
Hi @hengqujushi @sunchao, I rebased this and fixed the revapi failure, I think this should be ready for review now. Though I feel this PR is kind of big, maybe we should break this into smaller PRs for easier review? I think we can separate the column index filter part and record materialization on row read path part.  cc @rdblue What do you think?

iflytek-hmwang5 commented:
@zhongyujiang Hi, really appreciate the hard work has been done so far. It's a very useful feature! any update on this PR? Is there any work left to be done？

zhongyujiang commented:
Hi @iflytek-hmwang5 , We're waiting for people in the community who are interested in it to review it.

iflytek-hmwang5 commented:
> Hi @iflytek-hmwang5 , We're waiting for people in the community who are interested in it to review it.
What can we do to speed up the process? Looking forward to this feature very much.

zhongyujiang commented:
> @zhongyujiang Hi, what can do to speed up the process? Looking forward to this feature very much.

Um, I'm not sure, I think it depends on the community's priorities, and reviewers' bandwidth.

iflytek-hmwang5 commented:
> Um, I'm not sure, I think it depends on the community's priorities, and reviewers' bandwidth.
You can refer to this PR, https://github.com/apache/iceberg/issues/193, Anton Okolnychyi mentioned this idea before, maybe he'll be super interested.
